tests/kernels/test_self_and_cross_attn.py

```diff
@@ -0,0 +1,1250 @@
+
+    # Self-attention setup
+
+    self_block_base_addr = 0
+
+    query, \
+    prefill_packed_query, \
+    self_prefill_packed_key, \
+    self_prefill_packed_value, \
+    self_prefill_packed_ideal_output, \
+    prefill_q_seq_lens, \
+    self_prefill_kv_seq_lens, \
+    decode_packed_query, \
+    self_decode_packed_key, \
+    self_decode_packed_value, \
+    self_decode_packed_ideal_output, \
+    _, \
+    _, \
+    q_seq_lens, \
+    _, \
+    self_decode_block_tables, \
+    self_decode_slot_mapping, \
+    self_prefill_slot_mapping, \
+    self_prefill_block_tables, \
+    cross_block_base_addr = self_attn_setup(batch_size,
+                                                num_heads,
+                                                head_size,
+                                                block_size,
+                                                scale,
+                                                max_q_seq_len,
+                                                block_base_addr=self_block_base_addr)
+
+    # Cross-attention setup
+
+    cross_prefill_packed_key, \
+    cross_prefill_packed_value, \
+    cross_prefill_packed_ideal_output, \
+    cross_decode_packed_ideal_output, \
+    cross_kv_seq_lens, \
+    cross_decode_block_tables, \
+    cross_decode_slot_mapping, \
+    cross_prefill_slot_mapping, \
+    cross_prefill_block_tables, \
+    _ = cross_attn_setup_reuses_query(query,
+                                      q_seq_lens,
+                                      prefill_q_seq_lens,
+                                      batch_size,
+                                      num_heads,
+                                      head_size,
+                                      block_size,
+                                      scale,
+                                      max_q_seq_len,
+                                      max_kv_seq_len,
+                                      block_base_addr=cross_block_base_addr)
+
+    # PREFILL: self- and cross-attention tests
+
+    context_lens = [0 for _ in range(batch_size)]
+
+    prefill_attn_metadata: AttentionMetadata = make_metadata_self_cross(
+        attn_backend,
+        True,
+        prefill_q_seq_lens,
+        context_lens,
+        self_prefill_block_tables,
+        self_prefill_slot_mapping,
+        cross_seq_lens=cross_kv_seq_lens,
+        cross_block_tables=cross_prefill_block_tables,
+        cross_slot_mapping=cross_prefill_slot_mapping,
+    )
+
+    self_prefill_packed_actual_output: torch.Tensor = run_self_attention_test(
+        attn, prefill_packed_query, self_prefill_packed_key,
+        self_prefill_packed_value, kv_cache, prefill_attn_metadata)
+
+    # - Prefill self-attention correct?
+    assert torch.allclose(
+        self_prefill_packed_ideal_output,
+        self_prefill_packed_actual_output.view_as(
+            self_prefill_packed_ideal_output))
+
+    cross_prefill_packed_actual_output: torch.Tensor = run_cross_attention_test(
```

Comment:
Can you please clarify what we are trying the verify in this run_cross_attention_test? I was thinking we will have 3 things to verify here a) the encoder attention computation what we are doing and verifying right before this in L1202 (b) verify the decoder attention what we are doing in L1238 and c) verify the cross attention between the decoder attention and the encoder/cross attention what we are doing  L:1247.  So wondering which portion of attention computation this is verifying?

Reply:
Hi @sroy745 , there is a effectively a "grid" of test-case combinations. You are correct that one dimension of this grid is the (a)/(b)/(c) you have specified in your comment above, i.e. "type of attention".

However, another dimension of the test-case grid is prefill-phase vs decode-phase. This is because (1) prefill and decode use different kernels and (2) these kernels rely on different AttentionMetadata attributes being set & with different attribute values.

So the space of tests that needs to be considered is (prefill-phase, decode-phase) x (encoder test, decoder self-attention test, encoder/decoder cross-attention test); the only combination of test configurations which we skip is decode-phase + encoder test, because (under my implementation) the encoder inference is entirely completed during prefill, so there is no encoder inference during decode phase. Therefore there are 5 test conditions.

In order to make clearer how this is accomplished in my code, I recently made the following changes:
* run_self_attention_test() is renamed to run_encoder_or_decoder_self_attention_test(), so it is clear that this helper function may be used *either* to test encoder attention, or decoder self-attention (the attention metadata configuration is almost identical for both)
* run_cross_attention_test() is renamed to run_encoder_decoder_cross_attention_test(), so it is clear that this helper function may be used *only* to test encoder/decoder cross-attention
* I had actually forgotten to add a unit test for the prefill-phase & **encoder** test scenario; I just added it: test_encoder_attention()

If you look at test_enc_dec_self_and_cross_attention_prefill_decode_phases(), this test considers the following combination of scenarios:

(prefill-phase, decode-phase) x (decoder self-attention, encoder/decoder cross-attention)

The prefill-phase tests populate the self-attention and cross-attention KV caches, which are then re-used by the subsequent decode-phase tests.

The new encoder attention test I added, test_encoder_attention(), tests only the prefill-phase/encoder scenario. This scenario does not utilize the KV cache at all.

A note on terminology -

I have in the past used the term "cross-attention" interchangeably with "encoder/decoder attention", and I have used "self-attention" as if it were an adjective that describes both encoder attention and decoder self-attention (since both types of attention derive Q/K/V solely from outputs of the previous attention layer.) However based on your feedback, I think my terminology was possible unclear/incorrect. So I have adjusted to use the following terminology:

* The variety of attention that happens *inside* the encoder == "Encoder attention"
* The variety of attention where the decoder layer Q/K/V are all derived from the outputs of the previous decoder layer == "Decoder self-attention"
* The variety of attention where the decoder layer Q is derived from the previous layer output, while K & V are either derived from encoder hidden states or reused from cache == "Encoder/decoder cross attention"